lib/engine/game/g_18_mag/step/buy_train.rb

```diff
@@ -22,12 +22,23 @@ def buyable_trains(entity)
             # All trains start out available
             depot_trains = @depot.upcoming.group_by(&:name).map { |_k, v| v.first }
             other_trains = @depot.other_trains(entity)
-
-            depot_trains.reject! { |t| entity.cash < t.price }
+            depot_trains.reject! { |t| entity.cash < t.min_price(ability: train_discount_ability(entity)) }
             other_trains = [] if entity.cash < @game.class::TRAIN_PRICE_MIN
 
             depot_trains + other_trains
           end
+
+          def train_discount_ability(entity)
+            all_abilities = []
+            if entity.owner.respond_to?(:companies)
+              all_abilities += entity.owner.companies&.flat_map do |c|
```

Comment:
don't use +=, use concat, better yet though, don't use flat map, just do an each statement and insert it into all_aiblities